### ISSUE TITLE
[REQUEST] Better error handling.

### ISSUE THREAD

**Issue by i0bs** [interactions-py/interactions.py#692]
### Describe the feature.

Our current error handling is...
yeah, we agree too. It sucks.

I don't think we need to get into just how bad it is, but it's certainly a big pain for bot developers to deal with currently when using our library. Getting an ambiguous `50035` error code is confusing, and gives no details to the client on what is *actually wrong.* The DX for our error responses are subpar at best, and most of the time make no sense. On 23 March, the core development team discussed this matter internally and a proposed solution. On the 25th, we shared this information across both core and external development teams privately. Now, we will be publicizing the proposal which is still subject to feedback.

## Exceptions

Exceptions are a great way for bot developers to try a block of code, and if it fails, understand what went wrong with them. These are a standard practice in many applications which we have somewhat made use of with exceptions such as `InteractionException`, `HTTPException` and the sorts. However, these aren't versatile and flexible enough to support our needs.

Let's try that with **LibraryException**, our proposal for better exceptions.

```py
from typing import Optional
import logging

# We're going to subclass this off of the base Exception class from Python.
class LibraryException(Exception):
    code: Optional[int] # this is the error code of the exception which developers can extract.
    
    # this is the "severity" of the exception. this is something thrown as an idea to allow
    # developers to gather an understanding of just "how bad" an error may be. we're allowing
    # this to be an integer form to respect the logging level nature, as well as enable external
    # developers to use their own integer enumerables. this is a compromise between core lib
    # and 3rd party error exceptions.
    severity: int

    def __init__(self, message: str, code: int=None, severity: int=0):
        super().__init__(message=f"{message} (code {code} at {severity}.")
        self.code = code
        self.severity = severity
```

This allows us to try code, and raise exceptions which allow developers to not only customize but offer descriptive understanding for end-users.

```py
try:
  raise LibraryException("you're missing something important!", 
                         code=2, severity=logging.WARNING)
except LibraryException as error:
  print(error)
```

## Events

We've seen this as a common request from many bot developers to allow their applications to listen to errors as an event. These are a popular concept in libraries such as discord.py with `on_error`. We will be doing the same here, which will introduce:

- `on_command_error`
  - `on_command_option_error` (This is for things such as subcommands, technically considered options.)
- `on_component_error`
  - `on_select_option_error` (Select menu options have their own nested errors)
  - `on_button_error` (An abstract way of handling only button-related errors)
- `on_autocomplete_error`
- `on_modal_error`
  - `on_modal_option_error` (Modal options have their own nested errors)

When you listen to one of these events dispatched, you'll receive a `ctx` or `Context` variated class respective to the field. This means if your event listener is for command errors, you'll receive a `CommandContext` and vice versa.

```py
@bot.event
async def on_command_error(ctx: interactions.CommandContext, error):
    ...
```

In this case, `error` is a variation of many different exceptions. The *base* exception of the core library will be `LibraryException`, however, we're designing it with the intent of being subclassed. This opens up the door for us to rework `HTTPException` and other exceptions that we may find useful for bot developers.

## Error response

The response of an error is the biggest pain point for bot developers. I believe that it's a bit too early for us to get into how we're going to parse the response, but we can talk about how it'll appear for developers. We'll be providing a rigid structure to error reponses when one of the following meet:

- An HTTP request fails (excluding rate limits)
- The Gateway passes back an error

With this in mind, let's give an example. We'll be reading the JSON structure of [this error](https://discord.com/developers/docs/reference#error-messages-array-error) given from the Discord API docs.

```
{
    "code": 50035,
    "errors": {
        "activities": {
            "0": {
                "platform": {
                    "_errors": [
                        {
                            "code": "BASE_TYPE_CHOICES",
                            "message": "Value must be one of ('desktop', 'android', 'ios')."
                        }
                    ]
                },
                "type": {
                    "_errors": [
                        {
                            "code": "BASE_TYPE_CHOICES",
                            "message": "Value must be one of (0, 1, 2, 3, 4, 5)."
                        }
                    ]
                }
            }
        }
    },
    "message": "Invalid Form Body"
}
```

We'll see this error returned as this:

```
activities -> 0 -> platform -> Value must be one of ('desktop', 'android', 'ios') from base type choices. 
activities -> 0 -> platform -> Value must be one of (0, 1, 2, 3, 4, 5) from base type choices. 
(errors resulted from Invalid Form Body, code 50035)
```

Let's explain what this error response means. The first part, "activities" represents what resource/object in question we're having a problem with. The `0` represents this error as coming from an array, which the API will give us. From there, we get the attribute/field that is being problematic and then the message of that error itself. We will conjoin the error code alongside the message to add clarity to what the message means. If numerous errors are found, it'll list out all of the errors made. At the end, it'll tell us where the error or list of errors came from, and the general error code. We think that this is a huge improvement from the current DX of error handling.

This is an example of an error response we'd see from `HTTPException`, but can also be applicable to error responses we may see from places like the Gateway.

### Code of Conduct

- [X] I agree to follow the contribution requirements.

**Comment by i0bs** [comment]
Changes for this will now be directed to the branch `improved-error-handling`. A pull request will be made to signify when this is ready.

**Comment by sillyElvy** [comment]
This should be resolved by #848
